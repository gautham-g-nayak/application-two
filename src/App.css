.topSpace{
  padding-top: 3rem;
}