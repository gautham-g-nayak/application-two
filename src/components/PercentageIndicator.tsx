type PieProps = {
  numQuestions: number;
  total: number;
  color: string;
};

function PercentageIndicator({ numQuestions, total, color }: PieProps) {
  const percentage = (numQuestions / total) * 100;

  return (
    <div
      style={{
        display: "flex",
        justifyContent: "start",
        alignItems: "center",
        alignContent: "flex-start",
      }}
    >
      <div
        style={{
          background: color,
          border: "none",
          color: "white",
          padding: "10px 10px",
          textAlign: "center",
          textDecoration: "none",
          display: "inline-block",
          fontSize: "16px",
          margin: "10px 10px",
        }}
      >
        {percentage}%
      </div>
      <h4>- {color === "red" ? "incorrect answers" : "correct answers"}</h4>
    </div>
  );
}

export default PercentageIndicator;
