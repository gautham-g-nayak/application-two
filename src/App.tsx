import "./App.css";
import NavbarComponent from "./components/NavbarComponent";
import { BrowserRouter, Navigate, Route, Routes } from "react-router-dom";
import PageTwo from "./pages/PageTwo";
import PageThree from "./pages/PageThree";
import PageOne from "./pages/PageOne";

function App() {
  return (
    <BrowserRouter>
      <NavbarComponent />
      <div className="topSpace"></div>
      <Routes>
        <Route path="/one" element={<PageOne />} />
        <Route path="/two" element={<PageTwo />} />
        <Route path="/three" element={<PageThree />} />
        <Route path="*" element={<Navigate to="/one" />} />
      </Routes>
    </BrowserRouter>
  );
}

export default App;
