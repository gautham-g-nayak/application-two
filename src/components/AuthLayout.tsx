import { useContext } from "react";
import { Outlet, Navigate } from "react-router-dom";
import { AppContext } from "../App";

const AuthLayout = () => {
  const [appData] = useContext(AppContext);
  const length = Object.entries(appData.answers).length;
  return length === 0 ? (
    <Navigate to="/questions" replace />
  ) : (
    <Outlet />
  );
};

export default AuthLayout;
