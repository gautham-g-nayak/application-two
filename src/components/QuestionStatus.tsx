import { Box, Typography } from "@mui/material";
import questions from "../questions.json";
import { useContext } from "react";
import { AppContext } from "../App";

type QuestionStatusProps = {
    currentQuestion: number;
    setCurrentQuestion: React.Dispatch<React.SetStateAction<number>>;
  };

function QuestionStatus({ currentQuestion, setCurrentQuestion }: QuestionStatusProps) {
  const [appData, dispatchAppData] = useContext(AppContext);
  return (
    <Box
      width="340px"
      display="flex"
      flexWrap="wrap"
      flexDirection="row"
      justifyContent="start"
      marginLeft="20%"
      alignContent="flex-start"
    >
      <Typography variant="h6" padding="20px 10px">
        Question Status :
      </Typography>
      {questions.map((question, index) => (
        <button
          onClick={() => setCurrentQuestion(index + 1)}
          key={index}
          style={{
            background:
              currentQuestion === index + 1
                ? "#f5e72d"
                : appData.answers[index + 1]
                ? appData.answers[index + 1].value !== ""
                  ? "#4CAF50"
                  : "gray"
                : "gray",
            border: "none",
            borderRadius: "50px",
            color: "white",
            padding: "15px 20px",
            textAlign: "center",
            textDecoration: "none",
            display: "inline-block",
            fontSize: "16px",
            margin: "4px 4px",
            cursor: "pointer",
          }}
        >
          {index + 1}
        </button>
      ))}
    </Box>
  );
}

export default QuestionStatus;
