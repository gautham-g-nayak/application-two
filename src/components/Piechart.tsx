import PercentageIndicator from "./PercentageIndicator"

type PieProps = {
  correct: number;
  total: number;
};

function Piechart({ correct, total }: PieProps) {
  const correctDegree = (correct / total) * 360;
  const incorrectDegree = 360 - correctDegree;
  const pieStyle = () => ({
    height: "200px",
    width: "200px",
    backgroundColor: "green",
    backgroundImage: `conic-gradient(red ${incorrectDegree}deg, green ${incorrectDegree}deg ${correctDegree}deg)`,
    borderRadius: "50%",
  });
  return (
    <div>
      <div style={pieStyle()}></div>
      <PercentageIndicator numQuestions={correct} total={total} color={"green"}/>
      <PercentageIndicator numQuestions={total-correct} total={total} color={"red"}/>
    </div>
  );
}

export default Piechart;
