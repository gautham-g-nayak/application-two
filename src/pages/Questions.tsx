import { useState } from "react";
import QuestionCard, { QuestionType } from "../components/QuestionCard";
import questions from "../questions.json";
import { Box, Button, Typography } from "@mui/material";
import { useNavigate } from "react-router-dom";
import QuestionStatus from "../components/QuestionStatus";

export const TOTAL_QUESTIONS = questions.length;

const Questions = () => {
  const navigate = useNavigate();
  const [currentQuestion, setCurrentQuestion] = useState(questions[0].id);

  function handleNextButtonClick(index: number) {
    if (index === TOTAL_QUESTIONS - 1) {
      navigate("/results");
    } else {
      setCurrentQuestion((prev) => prev + 1);
    }
  }

  return (
    <Box
      height="100vh"
      display="flex"
      justifyContent="center"
      alignItems="center"
    >
      {questions.map(
        (question, index) =>
          currentQuestion === question.id && (
            <Box
              key={question.id}
              height="500px"
              display="flex"
              flexDirection="row"
              justifyContent="space-between"
            >
              <Box>
                <Typography variant="h3">Question: {index + 1}</Typography>
                <Box
                  height="500px"
                  display="flex"
                  flexDirection="column"
                  justifyContent="space-between"
                >
                  <QuestionCard
                    id={question.id}
                    question={question.question}
                    questionType={question.questionType as QuestionType}
                    answers={question.answerOptions}
                  />
                  <Box
                    sx={{
                      display: "flex",
                      alignItems: "center",
                      justifyContent: "space-between",
                    }}
                  >
                    <Button
                      disabled={index === 0}
                      onClick={() => setCurrentQuestion((prev) => prev - 1)}
                    >
                      Previous
                    </Button>

                    <Button onClick={() => handleNextButtonClick(index)}>
                      {index === TOTAL_QUESTIONS - 1 ? "Submit" : "Next"}
                    </Button>
                  </Box>
                </Box>
              </Box>
              <QuestionStatus
                currentQuestion={currentQuestion}
                setCurrentQuestion={setCurrentQuestion}
              />
            </Box>
          )
      )}
    </Box>
  );
};

export default Questions;
