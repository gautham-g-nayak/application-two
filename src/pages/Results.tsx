import { useContext, useEffect, useState } from "react";
import { AppContext } from "../App";
import questions from "../questions.json";
import { Box, Typography } from "@mui/material";
import Piechart from "../components/Piechart";

const Results = () => {
  const [appData] = useContext(AppContext);
  const [counter, setCounter] = useState(0);

  function correctOptions(index: number) {
    const options = questions[index].answerOptions.filter(
      (option) => option.isCorrect
    );
    const ans = options.map((option) => option.option);
    return ans;
  }

  function setValueOfPieChart() {
   
    let count = 0;
    Object.entries(appData.answers).map(([currentIndex, answer]: any) => {
      console.log(`appData : ${answer.value.toString()}`);
      console.log(`ans :${correctOptions(currentIndex - 1)}`);

      if (answer.value.toString() == correctOptions(currentIndex - 1)) {
        count++;
      }
    });
    setCounter(count);
  }

  useEffect(() => {
    setValueOfPieChart();
    const length = Object.entries(appData.answers).length;
    console.log(`length is ${length}`);
  }, [appData]);

  return (
    <Box
      height="100vh"
      display="flex"
      justifyContent="center"
      alignItems="center"
    >
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="space-between"
        width="500px"
        margin="10px 10px"
      >
        <Typography variant="h5" padding="5px">
          Report :
        </Typography>
        {Object.entries(appData.answers).map(([currentIndex, answer]: any) => (
          <div
            key={currentIndex}
            style={{
              padding: "5px",
            }}
          >
            <p
              style={{
                borderStyle: "solid",
                borderWidth: "2px",
                margin: "0px",
                padding: "10px 10px",
                borderRadius: "5px",
                borderColor:
                  answer.value.toString() == correctOptions(currentIndex - 1)
                    ? "green"
                    : "red",
              }}
            >
              {currentIndex}. {answer.value}
            </p>
            {answer.value.toString() == correctOptions(currentIndex - 1) ? (
              <p></p>
            ) : (
              <p>
                correct answer : {correctOptions(currentIndex - 1).join(", ")}
              </p>
            )}
          </div>
        ))}
      </Box>
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="center"
        alignItems="center"
        width="500px"
        margin="10px 10px"
      >
        <Piechart correct={counter}  total = {questions.length}/>
      </Box>
    </Box>
  );
};

export default Results;
