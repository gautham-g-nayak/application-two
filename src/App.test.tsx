import { fireEvent, render, screen } from "@testing-library/react";
import App, { AppContext } from "./App";
import { TOTAL_QUESTIONS } from "./pages/Questions";
import questions from "./questions.json";
import { actions } from "./actions";
import RadioButtons from "./components/buttons/RadioButtons";
import CheckboxOptions from "./components/CheckboxOptions";
import PercentageIndicator from "./components/PercentageIndicator";
import Piechart from "./components/Piechart";
import RadioOptions from "./components/RadioOptions";
import QuestionStatus from "./components/QuestionStatus";
import Results from "./pages/Results";
import QuestionCard from "./components/QuestionCard";
import AuthLayout from "./components/AuthLayout";
import {
  Box,
  Button,
  TextField,
  Paper,
  Typography,
  Grow,
  styled,
} from "@mui/material";

describe("unit testing basics", () => {
  test("renders Enter your details", () => {
    render(<App />);
    const element = screen.getByText(/Enter your details/i);
    expect(element).toBeInTheDocument();
  });

  test("renders Submit and Start Test", () => {
    render(<App />);
    const element = screen.getByText(/Submit and Start Test/i);
    expect(element).toBeInTheDocument();
  });

  test("renders result page", () => {
    const appData = {
      answers: {
        2: { type: "checkbox", value: ["string", "boolean"] },
      },
    };
    render(
      <AppContext.Provider value={[appData]}>
        <Results />
      </AppContext.Provider>
    );
    const element = screen.getByText(/Report :/i);
    expect(element).toBeInTheDocument();
  });

  test("renders Enter your details", () => {
    render(<App />);
    const element = screen.getByText(/Enter your details/i);
    expect(element).toBeInTheDocument();
  });

  test("assert if TOTAL_QUESTIONS has correct value", () => {
    expect(TOTAL_QUESTIONS).toBe(questions.length);
  });
});

describe("testing set answer", () => {
  test("testing set answer", () => {
    const { setAnswer } = actions;
    const newState = setAnswer(
      questions[0].id,
      questions[0].question,
      questions[0].questionType
    );

    const expectedState = {
      type: "SET_ANSWER",
      payload: {
        id: questions[0].id,
        data: {
          type: questions[0].questionType,
          value: questions[0].question,
        },
      },
    };
    expect(newState).toStrictEqual(expectedState);
  });
});

describe("testing components", () => {
  test("render radio buttons component", () => {
    render(
      <RadioButtons
        buttons={[
          { value: "male", label: "Male" },
          { value: "female", label: "Female" },
        ]}
        label=""
      />
    );
  });

  test("renders question card ", () => {
    const appData = {
      answers: {
        2: { type: "checkbox", value: ["string", "boolean"] },
      },
    };
    render(
      <AppContext.Provider value={[appData]}>
        <QuestionCard
          id={2}
          question={""}
          questionType={"checkbox"}
          answers={[]}
        />
      </AppContext.Provider>
    );
  });

  test("renders question card ", () => {
    const appData = {
      answers: {
        2: { type: "checkbox", value: ["string", "boolean"] },
      },
    };
    render(
      <AppContext.Provider value={[appData]}>
        <QuestionStatus currentQuestion={1} setCurrentQuestion={jest.fn()} />
      </AppContext.Provider>
    );
  });

  test("render check box options component", () => {
    const appData = {
      answers: {
        2: { type: "checkbox", value: ["string", "boolean"] },
      },
    };
    render(
      <CheckboxOptions
        options={questions[2].answerOptions.map((answer: any) => {
          return {
            option: answer.option,
            value: appData.answers[2]?.value.includes(answer.option) || false,
          };
        })}
        handleChange={jest.fn()}
      />
    );
  });

  test("render percentage indicator component", () => {
    const data = {
      numQuestions: 3,
      total: questions.length,
      color: "green",
    };
    render(
      <PercentageIndicator
        numQuestions={data.numQuestions}
        total={data.total}
        color={data.color}
      />
    );
  });

  test("render piechart component", () => {
    const data = {
      counter: 3,
    };
    render(<Piechart correct={data.counter} total={questions.length} />);
  });

  test("render radio options component", () => {
    const appData = {
      answers: {
        1: { type: "radio", value: "Every 15 days" },
      },
    };
    render(
      <RadioOptions
        options={questions[1].answerOptions.map((answer) => answer.option)}
        value={appData.answers[1]?.value || ""}
        handleChange={jest.fn()}
      />
    );
  });
});

describe("test Input Field component", () => {
  const StyledTextField = styled(TextField)({
    "& .MuiInputLabel-root ": {
      color: "#A9A9A9",
    },
    "& .MuiOutlinedInput-root": {
      "& fieldset": {
        borderColor: "inherit",
      },
    },
  });
  test("render Input Field component", () => {
    render(
      <StyledTextField
        fullWidth
        id="outlined-name"
        label="Name"
        variant="outlined"
      />
    );
    screen.debug();
  });

  test("snapshot testing", () => {
    const { container } = render(
      <StyledTextField
        fullWidth
        id="outlined-name"
        label="Name"
        variant="outlined"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
