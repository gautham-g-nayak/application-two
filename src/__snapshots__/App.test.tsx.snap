// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Input Field component snapshot testing 1`] = `
<div
  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1rqu6xr-MuiFormControl-root-MuiTextField-root"
>
  <label
    class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink="false"
    for="outlined-name"
    id="outlined-name-label"
  >
    Name
  </label>
  <div
    class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
  >
    <input
      aria-invalid="false"
      class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
      id="outlined-name"
      type="text"
      value=""
    />
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-1ftyaf0"
      >
        <span>
          Name
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
